src/Compute/Compute/VirtualMachine/Config/RemoveAzureVMDataDiskCommand.cs

```diff
@@ -53,20 +60,36 @@ public override void ExecuteCmdlet()
                     var disks = storageProfile.DataDisks.ToList();
                     var comp = StringComparison.OrdinalIgnoreCase;
 
-                    if (DataDiskNames == null)
-                    {
-                        disks.Clear();
+                    if (ForceDetach != true) {
+                        if (DataDiskNames == null){
+                            disks.Clear();
+                        }
+                        else{
+                            foreach (var diskName in DataDiskNames){
+                                disks.RemoveAll(d => string.Equals(d.Name, diskName, comp));
+                            }
+                        }
                     }
-                    else
-                    {
-                        foreach (var diskName in DataDiskNames)
+                    else{
+                        if (DataDiskNames == null){
```

Comment:
should we also add check for empty string/empty List?

Reply:
this is only for the case of empty list, string as this is a mandatory parameter